Errors due different devices when running Test locally.

When running `pytest --cov=clarity` locally on my laptop, some of the tests that use Pytorch fail due to different devices.
Note that I do have a GPU on my laptop.

The output error is:

> FAILED tests/regression/test_enhancers.py::test_dsp_filter - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cpu and cuda:0! (when checking argument for argum...
FAILED tests/regression/test_mc_conv_tasnet.py::test_convtasnet - RuntimeError: Input type (torch.cuda.FloatTensor) and weight type (torch.FloatTensor) should be the same

Maybe we need to be more explicit about the devices used in the test. We could create a util.device file, similar as proposed by @ns-rse in https://github.com/claritychallenge/clarity/issues/97#issuecomment-1313549828

--- a/clarity/enhancer/dnn/mc_conv_tasnet.py
+++ b/clarity/enhancer/dnn/mc_conv_tasnet.py
@@ -23,6 +23,7 @@ def overlap_and_add(signal, frame_step, device):
         signal: A [..., frames, frame_length] Tensor. All dimensions may be unknown,
             and rank must be at least 2.
         frame_step: An integer denoting overlap offsets. Must be less than or equal to frame_length.
+        device: Whether to use 'cpu' or 'cuda' for processing.
 
     Returns:
         A Tensor with shape [..., output_size] containing the overlap-added frames of
@@ -72,6 +73,7 @@ class ConvTasNet(nn.Module):
         norm_type="cLN",
         causal=False,
         mask_nonlinear="relu",
+        device=None,
     ):
         """
         Args:
@@ -113,7 +115,7 @@ class ConvTasNet(nn.Module):
         self.separator = TemporalConvNet(
             N_spec, N_spat, B, H, P, X, R, C, norm_type, causal, mask_nonlinear
         )
-        self.decoder = Decoder(N_spec, L)
+        self.decoder = Decoder(N_spec, L, device)
         # init
         for p in self.parameters():
             if p.dim() > 1:
@@ -192,10 +194,13 @@ class SpatialEncoder(nn.Module):
 
 
 class Decoder(nn.Module):
-    def __init__(self, N, L):
+    def __init__(self, N, L, device: str = None):
         super().__init__()
         # device for overlap_and add
-        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        if device is None:
+            self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        else:
+            self.device = device
         # Hyper-parameter
         self.N, self.L = N, L
         # Components

--- a/clarity/enhancer/dsp/filter.py
+++ b/clarity/enhancer/dsp/filter.py
@@ -7,9 +7,12 @@ EPS = 1e-8
 
 
 class AudiometricFIR(nn.Module):
-    def __init__(self, sr=44100, nfir=220):
+    def __init__(self, sr=44100, nfir=220, device=None):
         super().__init__()
-        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        if device is not None:
+            self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        else:
+            self.device = device
         self.window_size = nfir + 1
         self.padding = nfir // 2
 

--- a/clarity/predictor/torch_msbg.py
+++ b/clarity/predictor/torch_msbg.py
@@ -163,13 +163,17 @@ class MSBGHearingModel(nn.Module):
         spl_cali=True,
         src_posn="ff",
         kernel_size=1025,
+        device=None,
     ):
         super().__init__()
         self.sr = sr
         self.spl_cali = spl_cali
         self.src_posn = src_posn
         self.kernel_size = kernel_size
-        self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        if device is None:
+            self.device = "cuda" if torch.cuda.is_available() else "cpu"
+        else:
+            self.device = device
 
         # settings for audiogram
 

--- a/tests/regression/test_enhancers.py
+++ b/tests/regression/test_enhancers.py
@@ -24,7 +24,7 @@ gha_params = {  # hyperparameters for GHA Hearing Aid, BE CAREFUL if making chan
 
 
 def test_dsp_filter(regtest):
-    amfir = filter.AudiometricFIR(sr=44100, nfir=220)
+    amfir = filter.AudiometricFIR(sr=44100, nfir=220, device="cpu")
     torch.manual_seed(0)
     signal = torch.rand(10, dtype=torch.float)
     signal = torch.reshape(signal, (1, 1, -1))

--- a/tests/regression/test_mc_conv_tasnet.py
+++ b/tests/regression/test_mc_conv_tasnet.py
@@ -35,6 +35,7 @@ def test_convtasnet(regtest):
         "norm_type": "cLN",
         "causal": True,
         "mask_nonlinear": "relu",
+        "device": "cpu",
     }
 
     cfg["test_dataset"] = {

--- a/tests/regression/test_predictors.py
+++ b/tests/regression/test_predictors.py
@@ -15,11 +15,9 @@ def test_torch_msbg_stoi(regtest):
 
     audiogram = [45, 35, 30, 45, 50, 50]
     audiometric = [250, 500, 1000, 2000, 4000, 6000]
-    msbg = MSBGHearingModel(audiogram=audiogram, audiometric=audiometric)
+    msbg = MSBGHearingModel(audiogram=audiogram, audiometric=audiometric, device="cpu")
 
     x = torch.randn(2, 44100)
-    if torch.cuda.is_available():
-        x = x.cuda()
     y = msbg(x)
     stoi_loss = stoi_loss.forward(x.cpu(), y.cpu()).mean()
     estoi_loss = estoi_loss.forward(x.cpu(), y.cpu()).mean()
@@ -36,8 +34,7 @@ def test_torchloudnorm(regtest):
     ln = torchloudnorm()
 
     x = torch.randn(2, 44100)
-    if torch.cuda.is_available():
-        x = x.cuda()
+    x = x.cpu()
     y = ln(x)
     div = (y / (x + 1e-8)).cpu().mean()
 

